cpp/include/cudf/io/readers.hpp

```diff
@@ -341,6 +341,20 @@ class reader {
   std::unique_ptr<impl> _impl;
 
  public:
+  /**
+   * @brief Constructor for external datasource.
+   *
+   * @param external_datasource_id Unique identifier of the externally loaded datasource
+   * @param datasource_confs Map of <string, string> key/value pairs that will be used in
+   * configuring the external datasource
+   * @param options Settings for controlling reading behavior
+   * @param mr Optional resource to use for device memory allocation
+   */
+  explicit reader(std::string external_datasource_id,
```

Comment:
I think it would be better to pass std::unique_ptr<cudf::io::datasource> to the constructor. This way the logic of the datasource creation (which could be Kafka) is kept outside of the reader. Avoid duplication once multiple reader support this source type. We have this setup already implemented for data_sink and the writers, you can implement the equivalent here.